### PR TITLE
Job constraints to Kube primitives mapping

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/JobConstraints.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+package com.netflix.titus.api.jobmanager;
+
+import java.util.Set;
+
+import static com.netflix.titus.common.util.CollectionsExt.asSet;
+
+/**
+ * Constants and functions for job soft and hard constraints. Constraint names are case insensitive.
```
Comment (fabiokung) [inline titus-api/src/main/java/com/netflix/titus/api/jobmanager/JobConstraints.java]: it seems `SchedulingConstaintValidator` assumes they are always lowercase

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultTaskToPodConverter.java [hunk 2]
```diff
@@ -114,7 +122,8 @@ public V1Pod apply(Job<?> job, Task task) {
                 .schedulerName("default-scheduler")
                 .containers(Collections.singletonList(container))
                 .terminationGracePeriodSeconds(POD_TERMINATION_GRACE_PERIOD_SECONDS)
-                .restartPolicy(NEVER_RESTART_POLICY);
+                .restartPolicy(NEVER_RESTART_POLICY)
+                .affinity(podAffinityFactory.buildV1Affinity(job, task));
```
Comment (fabiokung) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultTaskToPodConverter.java]: should we add this only if `isOwnedByKubeScheduler()`? Any harm in adding it when tasks are being scheduled by Fenzo? (i.e.: will it negatively impact other controllers?)

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/JobConstraints.java (hunk 1 above)
Comment (tbak) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/JobConstraints.java]: We ignore case, so it is ok for a user to pass for example `ExclusiveHost`

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultTaskToPodConverter.java (hunk 2 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultTaskToPodConverter.java]: This check is made earlier. This code is executed only with direct (non Fenzo) scheduling.

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            this.job = job;
+            this.task = task;
+            this.v1Affinity = new V1Affinity();
+
+            processJobConstraints(job.getJobDescriptor().getContainer().getHardConstraints(), true);
+            processJobConstraints(job.getJobDescriptor().getContainer().getSoftConstraints(), false);
+        }
+
+        private void processJobConstraints(Map<String, String> constraints, boolean hard) {
+            if ("true".equalsIgnoreCase(constraints.get(JobConstraints.EXCLUSIVE_HOST))) {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java]: Boolean.valueOf?

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                addExclusiveHostConstraint(hard);
+            }
+
+            if ("true".equalsIgnoreCase(constraints.get(JobConstraints.UNIQUE_HOST))) {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java]: Boolean.valueOf?

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/PodAffinityFactory.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package com.netflix.titus.master.mesos.kubeapiserver.direct;
+
+import com.netflix.titus.api.jobmanager.model.job.Job;
+import com.netflix.titus.api.jobmanager.model.job.Task;
+import io.kubernetes.client.models.V1Affinity;
+
+public interface PodAffinityFactory {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/PodAffinityFactory.java]: Can you add javadocs here?

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java (hunk 3 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java]: Yes. Even better with `Boolean.parseBoolean`

**Pull request merged**